src/lib.rs

```diff
@@ -74,51 +75,115 @@ impl Notifier {
-        )
+    if let Some(v) = map.remove("azurite_host") {
+        let mut azurite_host = url::Url::parse(&v)
+            .map_err(|e| anyhow!("failed to parse azurite_host: {}", e))?;
+        azurite_host.set_path("");
+        std::env::set_var("AZURITE_BLOB_STORAGE_URL", azurite_host.as_str());
+        map.insert("allow_invalid_certificates".into(), "true".into());
+        map.insert("azure_storage_use_emulator".into(), "true".into());
     }
+
+    if let Some(v) = map.remove("minio_host") {
+        let mut minio_host = url::Url::parse(&v)
+            .map_err(|e| anyhow!("failed to parse minio_host: {}", e))?;
+        minio_host.set_path("");
+        // std::env::set_var("AZURITE_BLOB_STORAGE_URL", azurite_host.as_str());
+        map.insert("allow_http".into(), "true".into());
+        map.insert("aws_endpoint".into(), minio_host.as_str().trim_end_matches('/').to_string());
+        tracing::warn!("aws endpoint: {}", map["aws_endpoint"]);
+    }
+
+    let mut retry_config = RetryConfig::default();
+    if let Some(value) = map.remove("max_retries") {
+        retry_config.max_retries = value.parse()
+            .map_err(|e| anyhow!("failed to parse max_retries: {}", e))?;
+    }
+    if let Some(value) = map.remove("retry_timeout_secs") {
+        retry_config.retry_timeout = std::time::Duration::from_secs(value.parse()
+            .map_err(|e| anyhow!("failed to parse retry_timeout_sec: {}", e))?
+        );
+    }
+
+    let client: Arc<dyn ObjectStore> = match url.scheme() {
+        "s3" => {
+            let mut builder = object_store::aws::AmazonS3Builder::default()
+                .with_url(url)
+                .with_retry(retry_config);
+            for (key, value) in map {
+                builder = builder.with_config(key.parse()?, value);
+            }
+            Arc::new(builder.build()?)
+        }
+        "az" | "azure" => {
+            let mut builder = object_store::azure::MicrosoftAzureBuilder::default()
+                .with_url(url)
+                .with_retry(retry_config);
+            for (key, value) in map {
+                builder = builder.with_config(key.parse()?, value);
+            }
+            Arc::new(builder.build()?)
+        }
+        _ => unimplemented!("unknown url scheme")
+    };
+    // let (store, _) = object_store::parse_url_opts(&url, pairs)?;
+
+    // let client: Arc<dyn ObjectStore> = Arc::new(store);
+
+    // let ping_path: Path = "_this_file_does_not_exist".try_into().unwrap();
+    // match client.get(&ping_path).await {
+    //     Ok(_) | Err(object_store::Error::NotFound { .. }) => {},
+    //     Err(e) => {
+    //         return Err(anyhow!("failed to check store client connection: {}", e));
+    //     }
+    // }
+
+    Ok(client)
 }
 
-unsafe impl Send for AzureCredentials {}
-unsafe impl Sync for AzureCredentials {}
+#[repr(C)]
+pub struct GlobalConfigOptions {
+    max_retries: usize,
+    retry_timeout_sec: u64,
```

Comment:
Also, are we planning to drop the `GlobalConfigOptions` in future and make all config request-specific? (Or, as i asked above, are these defaults that we use if the request does set these values?)

Reply:
I was planning on dropping it to bring it back later when we truly need some config at initialization, but I realized we could use it for the `n_threads` config and changed it to that